### PR TITLE
addmv decomp #2

### LINKED ISSUE #94617: Write a decomposition for addmv
### 🐛 Describe the bug

The operator is

```
- func: addmv.out(Tensor self, Tensor mat, Tensor vec, *, Scalar beta=1, Scalar alpha=1, Tensor(a!) out) -> Tensor(a!)
  structured: True
  dispatch:
    CPU: addmv_out_cpu
    CUDA: addmv_out_cuda
    MPS: addmv_out_mps
    SparseCsrCPU: addmv_out_sparse_compressed
    SparseCsrCUDA: addmv_out_sparse_compressed_cuda
```

Look in `torch/_decomp/decompositions.py` for examples. You can decompose it according to the formula in the documentation. However, this may not correctly handle stride behavior; talk to your bootcamp mentor if it does not.

It may be necessary to ensure this decomp is disabled for inductor (which currently does not compile matrix multiplies).

Once you're done python test/test_proxy_tensor.py -k addmv should xpass. Remove the xfail for addmv. Other tests may xpass too

### Versions

master

cc @SherlockNoMad @ngimel

### PR DESCRIPTION
**Pull request by cybershiptrooper**
Fixes #94617

### CODE REVIEW

**Comment by cybershiptrooper** [comment]
@ngimel could you please review this again?
I messed up merging the previous one 
(again, sorry about that: pretty new to open source..)
Thanks a lot

**Review by ngimel — APPROVED** [review]

**Comment by ezyang** [comment]
Test failures are real 🤔

**Comment by cybershiptrooper** [comment]
I might need some help here :(
The accuracy error does not show up on replacing `torch.mv` with `@/matmul/(A*b).sum(dim=1)`, but I couldn't understand why.. 

Even if I print something like this without changing anything, it fails: 
```
out = alpha * torch.matmul(mat1, vec)
print(torch.linalg.norm( (mat1 * vec).sum(dim=1) - (torch.mv(mat1, vec)) ))
return out + beta * self
``` 
(=8.5299e-06)

And I saw [this](https://github.com/pytorch/pytorch/blob/fe05266fda4f908130dea7cbac37e9264c0429a2/test/test_decomp.py#L210) but removing it doesn't affect any mv related tests, so idk why/when did mv itself faced accuracy issues, or if this changes addmv's behaviour(but why would printing something random cause problem?)

Also, I couldn't understand the `test_has_decomposition` error: shouldn't `@register_decomposition` add it to the table?

**Comment by ngimel** [comment]
I guess you need to add higher tolerance for addmv to that table

**Comment by ezyang** [comment]
Is the precision difference due to FMA or something?

**Comment by ngimel** [comment]
blas gemv likely compiles to fma, while `out + beta * self` doesn't

**Comment by ezyang** [comment]
I wonder if we should have an FMA prim for this reason 🤔

**Comment by cybershiptrooper** [comment]
@ngimel I just have one more question for clarity: matmul just forwards the call to torch.mv (in this case). So shouldn't matmul also use the same rounding? Why would matmul pass and not mv if they both call the same function? The print thing also went over my head 
Thanks a lot

**Comment by ngimel** [comment]
Yes, matmul should call mv. Sorry, I don't quite understand what fails and what passes from your comments, can you post a standalone runnable script demonstrating the differences between matmul and mv?

**Comment by cybershiptrooper** [comment]
There isn't any difference in the result but somehow, matmul passes the accuracy test on my system.. ( instead of using torch.mv(mat1, vec) )

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -1119,7 +1119,7 @@ def split(self: Tensor, split_size: int, dim: int = 0) -> List[Tensor]:
 @register_decomposition(aten.addmm)
 @out_wrapper()
 @pw_cast_for_opmath
-def addmm(self: Tensor, mat1: Tensor, mat2: Tensor, beta: int = 1, alpha: int = 1):
+def addmm(self: Tensor, mat1: Tensor, mat2: Tensor, beta=1, alpha=1):
```
Comment (lezcano) [inline torch/_decomp/decompositions.py:1122]: This change is unrelated to this PR.

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -1136,6 +1136,19 @@ def addmm(self: Tensor, mat1: Tensor, mat2: Tensor, beta: int = 1, alpha: int =
     return out + beta * self
 
 
+@register_decomposition(aten.addmv)
+@out_wrapper()
+@pw_cast_for_opmath
+def addmv(self: Tensor, mat1: Tensor, vec: Tensor, beta=1, alpha=1):
```
Comment (lezcano) [inline torch/_decomp/decompositions.py:1142]: Let's annotate `beta, alpha` now that we're at it.

File: torch/_decomp/decompositions.py [hunk 3]
```diff
@@ -1136,6 +1136,19 @@ def addmm(self: Tensor, mat1: Tensor, mat2: Tensor, beta: int = 1, alpha: int =
...
+    if not self.is_floating_point() and not self.is_complex():
+        beta = int(beta)
+        alpha = int(alpha)
+    out = alpha * torch.matmul(mat1, vec)
```
Comment (lezcano) [inline torch/_decomp/decompositions.py:1146]: I think we should decompose it to `torch.mv`, as we are doing with `torch.addmm`. I wouldn't worry about some minor accuracy issues, which may very well be coming from the use of FMA, as Natalia and Ed described.

About whether to add an FMA prim or not, it may actually make sense. One of the design principles for PrimTorch was that anything that had its own function on the STL would be a prim. This makes sense, as the STL chooses these considering their speed / numerical stability. In this case, it looks like FMA may fit both speed and stability so well worth having it IMO. Perhaps, in this case, this could be not a prim but a decomposition, and then backends could choose to override it with their own implementation?

**Comment by ngimel** [comment]
Sorry, both things can't be true
1) there's no difference in the results with matmul and mv
2) matmul passes the test and mv doesn't. 
Can you please dig deeper and see what's going on? 
Or leave `mv` in decomposition and increase tolerance for the tests.

**Comment by cybershiptrooper** [comment]
@ngimel 
Whenever I go through matmul, both orig and decomp(in DecompCrossRefMode) have low accuracy(~1e-2 to ~1e-4), and mv's test passes(holds for all seeds).. Not sure what is happening yet. If there is something wrong, I'll add an issue separately.
Error in mv's test is due to (self * vec) not using FMA in its decomp(increased its tolerance). No need to increase addmv's tolerance. 
I hope this is fine. Am I allowed to merge this using pytorchbot? Or do I need all reviews to do so?
Thanks

**Comment by ngimel** [comment]
I don't understand how `orig` can have different accuracy depending on decomposition, but you can land PR in its current form if the tests pass.

**Comment by cybershiptrooper** [comment]
ahh.. Could you please run again? Sorry for the trouble

**Review by ezyang — APPROVED** [review]
